File: packages/python-sdk/rasmcore/pipeline.py
"""WASM Component Model pipeline — loads rasmcore-image.wasm and exposes Python API."""

from __future__ import annotations

import os
from pathlib import Path
from typing import Optional

import wasmtime
from wasmtime import Config, Engine, Linker, Module, Store


def _find_wasm() -> Path:
    """Locate the rasmcore-image WASM component binary."""
    # Check common locations
    candidates = [
        Path(__file__).parent / "rasmcore_image.wasm",
        Path.cwd() / "target" / "wasm32-wasip1" / "release" / "rasmcore_image.wasm",
        Path.cwd() / "target" / "wasm32-wasip1" / "debug" / "rasmcore_image.wasm",
    ]
    env_path = os.environ.get("RASMCORE_WASM_PATH")
    if env_path:
        candidates.insert(0, Path(env_path))

    for p in candidates:
        if p.exists():
            return p
    raise FileNotFoundError(
        "rasmcore_image.wasm not found. Set RASMCORE_WASM_PATH or build with: "
        "cargo component build -p rasmcore-image --release"
    )


class Pipeline:
    """Image processing pipeline backed by WASM Component Model.

    Usage:
        from rasmcore import Pipeline

        pipe = Pipeline()
        with open("input.jpg", "rb") as f:
            img = pipe.read(f.read())
        blurred = pipe.blur(img, radius=5.0)
        result = pipe.write_png(blurred)
        with open("output.png", "wb") as f:
            f.write(result)
    """

    def __init__(
        self,
        wasm_path: Optional[str] = None,
        use_gpu: bool = True,
    ):
        self._use_gpu = use_gpu
        self._gpu_executor = None

        # Initialize GPU if requested
        if use_gpu:
            try:
                from .gpu_host import WgpuExecutor
                self._gpu_executor = WgpuExecutor()
            except Exception:
                pass  # GPU unavailable — CPU fallback

        # Load WASM component
        path = Path(wasm_path) if wasm_path else _find_wasm()

        config = Config()
        config.wasm_component_model = True
        self._engine = Engine(config)
        self._linker = Linker(self._engine)
        self._linker.define_wasi()

        # If GPU available, provide the gpu-execute import
        if self._gpu_executor:
            self._register_gpu_import()

        self._module = Module.from_file(self._engine, str(path))
        self._store = Store(self._engine)
        self._instance = self._linker.instantiate(self._store, self._module)

        # Get pipeline resource constructor
        self._pipeline_new = self._instance.exports(self._store).get("pipeline", "new")
        self._pipeline = None

    def _register_gpu_import(self):
        """Register the gpu-execute host function for GPU offload."""
        executor = self._gpu_executor

        def gpu_execute(ops, input_bytes, width, height):
            """Host-side GPU execution via wgpu."""
            from .gpu_host import translate_ops
            native_ops = translate_ops(ops)
            return executor.execute(native_ops, input_bytes, width, height)

        # Register as WIT import
        # Note: exact registration depends on wasmtime-py component model API
        # This is a placeholder — the actual binding depends on the generated types
        self._gpu_execute_fn = gpu_execute

    def _ensure_pipeline(self):
        if self._pipeline is None:
            self._pipeline = self._pipeline_new(self._store)

    def read(self, image_bytes: bytes) -> int:
        """Read encoded image bytes (PNG, JPEG, etc.) and return a node ID."""
        self._ensure_pipeline()
        read_fn = self._instance.exports(self._store).get("pipeline", "read")
        return read_fn(self._store, self._pipeline, image_bytes)

    def blur(self, source: int, radius: float = 5.0) -> int:
        """Apply gaussian blur."""
        self._ensure_pipeline()
        blur_fn = self._instance.exports(self._store).get("pipeline", "blur")
        return blur_fn(self._store, self._pipeline, source, {"radius": radius})

    def spherize(self, source: int, strength: float = 0.5) -> int:
        """Apply spherize distortion."""
        self._ensure_pipeline()
        fn = self._instance.exports(self._store).get("pipeline", "spherize")
        return fn(self._store, self._pipeline, source, {"strength": strength})

    def sepia(self, source: int, intensity: float = 1.0) -> int:
        """Apply sepia tone."""
        self._ensure_pipeline()
        fn = self._instance.exports(self._store).get("pipeline", "sepia")
        return fn(self._store, self._pipeline, source, {"intensity": intensity})

    def hue_rotate(self, source: int, degrees: float = 90.0) -> int:
        """Rotate hue."""
        self._ensure_pipeline()
        fn = self._instance.exports(self._store).get("pipeline", "hue-rotate")
        return fn(self._store, self._pipeline, source, {"degrees": degrees})

    def write_png(self, source: int, gpu: Optional[bool] = None) -> bytes:
        """Encode as PNG and return bytes."""
        self._ensure_pipeline()
        write_fn = self._instance.exports(self._store).get("pipeline", "write-png")
        config = {}
        return write_fn(self._store, self._pipeline, source, config, None)

    def write_jpeg(self, source: int, quality: int = 85, gpu: Optional[bool] = None) -> bytes:
        """Encode as JPEG and return bytes."""
        self._ensure_pipeline()
        write_fn = self._instance.exports(self._store).get("pipeline", "write-jpeg")
        config = {"quality": quality}
        return write_fn(self._store, self._pipeline, source, config, None)

    def write(self, source: int, format: str = "png", quality: int = 85, gpu: Optional[bool] = None) -> bytes:
        """Generic write — dispatches to format-specific encoder."""
        self._ensure_pipeline()
        write_fn = self._instance.exports(self._store).get("pipeline", "write")
        return write_fn(
            self._store, self._pipeline, source,
            format, quality if quality > 0 else None, None,
        )

File: packages/go-sdk/rasmcore/pipeline.go
// Package rasmcore provides a Go SDK for the rasmcore WASM image processing pipeline.
//
// The SDK loads the rasmcore-image WASM component via wazero (pure Go, no CGO)
// and exposes a Pipeline type for image operations. Optional GPU acceleration
// is available via the gpu sub-package (requires wgpu-native CGO bindings).
//
// Usage:
//
//	pipe, err := rasmcore.NewPipeline(rasmcore.Options{})
//	if err != nil { log.Fatal(err) }
//	defer pipe.Close()
//
//	img, err := pipe.Read(imageBytes)
//	if err != nil { log.Fatal(err) }
//
//	blurred, err := pipe.Blur(img, rasmcore.BlurConfig{Radius: 10.0})
//	if err != nil { log.Fatal(err) }
//
//	result, err := pipe.WriteJPEG(blurred, rasmcore.JPEGConfig{Quality: 90})
//	if err != nil { log.Fatal(err) }
package rasmcore

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	"github.com/tetratelabs/wazero"
	"github.com/tetratelabs/wazero/api"
	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
)

// Options configures the Pipeline.
type Options struct {
	// WASMPath overrides the default WASM component location.
	// If empty, searches standard locations and RASMCORE_WASM_PATH env var.
	WASMPath string

	// UseGPU enables GPU acceleration if available. Default: true.
	UseGPU bool

	// GpuExecutor provides the GPU execution implementation.
	// If nil and UseGPU is true, attempts auto-detection.
	GpuExecutor GpuExecutor
}

// GpuExecutor is the interface for GPU compute offload.
// Implement this to provide GPU support (see gpu sub-package for wgpu-native impl).
type GpuExecutor interface {
	// Execute runs a batch of GPU compute operations on pixel data.
	// ops are chained: output of ops[i] = input of ops[i+1].
	Execute(ops []GpuOp, input []byte, width, height uint32) ([]byte, error)

	// MaxBufferSize returns the maximum buffer size in bytes.
	MaxBufferSize() int

	// Close releases GPU resources.
	Close()
}

// GpuOp represents a single GPU compute dispatch.
type GpuOp struct {
	Shader       string   // WGSL compute shader source
	EntryPoint   string   // Shader entry point name
	WorkgroupX   uint32   // Workgroup size X
	WorkgroupY   uint32   // Workgroup size Y
	WorkgroupZ   uint32   // Workgroup size Z
	Params       []byte   // Uniform parameters
	ExtraBuffers [][]byte // Additional storage buffers
}

// BlurConfig configures gaussian blur.
type BlurConfig struct {
	Radius float32
}

// SpherizeConfig configures spherize distortion.
type SpherizeConfig struct {
	Strength float32
}

// SepiaConfig configures sepia tone.
type SepiaConfig struct {
	Intensity float32
}

// JPEGConfig configures JPEG encoding.
type JPEGConfig struct {
	Quality uint8
	GPU     *bool // nil = use default, false = force CPU
}

// PNGConfig configures PNG encoding.
type PNGConfig struct {
	GPU *bool
}

// NodeID is a handle to a pipeline node.
type NodeID = uint32

// Pipeline wraps the rasmcore WASM image processing component.
type Pipeline struct {
	runtime wazero.Runtime
	module  api.Module
	ctx     context.Context
	gpu     GpuExecutor
}

func findWASM(override string) ([]byte, error) {
	candidates := []string{
		override,
		os.Getenv("RASMCORE_WASM_PATH"),
		filepath.Join(".", "rasmcore_image.wasm"),
		filepath.Join("target", "wasm32-wasip1", "release", "rasmcore_image.wasm"),
		filepath.Join("target", "wasm32-wasip1", "debug", "rasmcore_image.wasm"),
	}
	for _, p := range candidates {
		if p == "" {
			continue
		}
		data, err := os.ReadFile(p)
		if err == nil {
			return data, nil
		}
	}
	return nil, fmt.Errorf(
		"rasmcore_image.wasm not found; set RASMCORE_WASM_PATH or build with: " +
			"cargo component build -p rasmcore-image --release",
	)
}

// NewPipeline creates a new image processing pipeline.
//
// The WASM component is loaded once and reused for all operations.
// Call Close() when done to release resources.
func NewPipeline(opts Options) (*Pipeline, error) {
	ctx := context.Background()

	wasmBytes, err := findWASM(opts.WASMPath)
	if err != nil {
		return nil, err
	}

	rt := wazero.NewRuntime(ctx)

	// Instantiate WASI
	wasi_snapshot_preview1.MustInstantiate(ctx, rt)

	// TODO: Register gpu-execute host function if GPU available
	var gpu GpuExecutor
	if opts.UseGPU && opts.GpuExecutor != nil {
		gpu = opts.GpuExecutor
		// Register host function for gpu-execute WIT import
		// This requires component model support in wazero
	}

	// Compile and instantiate
	compiled, err := rt.CompileModule(ctx, wasmBytes)
	if err != nil {
		rt.Close(ctx)
		return nil, fmt.Errorf("compile WASM: %w", err)
	}

	mod, err := rt.InstantiateModule(ctx, compiled, wazero.NewModuleConfig())
	if err != nil {
		rt.Close(ctx)
		return nil, fmt.Errorf("instantiate WASM: %w", err)
	}

	return &Pipeline{
		runtime: rt,
		module:  mod,
		ctx:     ctx,
		gpu:     gpu,
	}, nil
}

// Close releases all pipeline resources.
func (p *Pipeline) Close() error {
	if p.gpu != nil {
		p.gpu.Close()
	}
	return p.runtime.Close(p.ctx)
}

// Read decodes image bytes (PNG, JPEG, etc.) into a pipeline node.
func (p *Pipeline) Read(imageBytes []byte) (NodeID, error) {
	// TODO: Call WASM pipeline.read() via component model
	// wazero doesn't support Component Model yet — this is a placeholder
	// for when wazero adds CM support or we use wasmtime-go instead
	return 0, fmt.Errorf("WASM Component Model not yet supported in wazero; " +
		"use wasmtime-go or wait for wazero CM support")
}

// Blur applies gaussian blur to a node.
func (p *Pipeline) Blur(source NodeID, config BlurConfig) (NodeID, error) {
	return 0, fmt.Errorf("not yet implemented — pending CM support")
}

// Spherize applies spherize distortion.
func (p *Pipeline) Spherize(source NodeID, config SpherizeConfig) (NodeID, error) {
	return 0, fmt.Errorf("not yet implemented — pending CM support")
}

// Sepia applies sepia tone.
func (p *Pipeline) Sepia(source NodeID, config SepiaConfig) (NodeID, error) {
	return 0, fmt.Errorf("not yet implemented — pending CM support")
}

// WriteJPEG encodes a node as JPEG.
func (p *Pipeline) WriteJPEG(source NodeID, config JPEGConfig) ([]byte, error) {
	return nil, fmt.Errorf("not yet implemented — pending CM support")
}

// WritePNG encodes a node as PNG.
func (p *Pipeline) WritePNG(source NodeID, config PNGConfig) ([]byte, error) {
	return nil, fmt.Errorf("not yet implemented — pending CM support")
}
